### PR TITLE
gh-116818: Make `sys.settrace`, `sys.setprofile`, and monitoring thread-safe

### PR DESCRIPTION
**Pull request by DinoV**
Stops the world when we're doing a monitoring event that impacts all threads.  

Updates the initialization of the monitoring data structures so that modifications generally lock the code object.  

There's various side data structures depending upon the tracing modes that are enabled.  These are all allocated and published once.  This is necessary because `_Py_Instrument` can be called on code objects pretty freely and we probably don't want locks on the readers.  Races with the actual values should be benign and just result in some delay in seeing the events being enabled/disabled.

Also adds multi-threaded test cases which are passing with `PYTHON_GIL=0`.

* Issue: gh-116818

### CODE REVIEW

File: Python/instrumentation.c
Comment (colesbury) [inline Python/instrumentation.c]:
```suggestion
get_monitoring(PyCodeObject *code)
{
```

File: Python/legacy_tracing.c
Comment (colesbury) [inline Python/legacy_tracing.c]:
```suggestion
    Py_ssize_t profiling_threads = setup_profile(tstate, func, arg, &old_profileobj);
```

File: Python/instrumentation.c
Comment (colesbury) [inline Python/instrumentation.c]: I don't think `PyErr_Format` is safe to call during a stop-the-world pause -- it can be moved immediate after `_PyEval_StartTheWorld()`.

File: Python/instrumentation.c
Comment (colesbury) [inline Python/instrumentation.c:2267]: Some parts of `instrument_all_executing_code_objects` are not safe in a stop-the-world pause -- it's not safe to call `Py_DECREF` or other functions that may execute Python code.

**Review by colesbury — COMMENTED** [review]
It's not clear to me that the strategy of *either* locking the code object or a stop-the-world pause is safe.

* Is locking code objects sufficient? What if other threads are executing those code objects?
* We need to be careful about what we do doing a stop-the-world pause. It's not safe to call `Py_DECREF` either directly or indirectly, for example.

File: Include/internal/pycore_pyatomic_ft_wrappers.h
Comment (markshannon) [inline Include/internal/pycore_pyatomic_ft_wrappers.h:29]: Wrapping things in macros often obfuscates the intent.
If I see `_Py_atomic_load_ptr_acquire(&value)` I know what it does (assuming I know how atomics work)
If I see `FT_ATOMIC_LOAD_PTR_ACQUIRE(value)` I need to scan the source code to find the macro definition.

Can we come up with a name that describes the semantics, not the context in which it should be used?
All these operations are thread safe on all platforms. Either the GIL makes them safe or we use hardware atomics.
So, maybe choose a name based on the semantics not the context, and drop the `FT_` prefix?

OOI, do we have any evidence that relaxed/release/acquire gains us performance over sequential consistency which is a lot easier to reason about and thus safer.

File: Python/instrumentation.c
Comment (markshannon) [inline Python/instrumentation.c]: I think it would be better to assert that this is locked here and move the critical section to the `DISABLE` section of `call_instrumentation_vector`. `_Py_Instrument` should already be locked (with stop the world)

File: Python/instrumentation.c
Comment (markshannon) [inline Python/instrumentation.c]: No need for atomic operations here. `per_instruction_opcodes` is updated in `_Py_Instrument` which should only be called with the world stopped, so it cannot change while threads are running.

File: Python/instrumentation.c
Comment (markshannon) [inline Python/instrumentation.c]: Likewise

File: Python/instrumentation.c
Comment (markshannon) [inline Python/instrumentation.c]: Maybe drop the `#ifdef Py_GIL_DISABLED` and use `(FT_)ATOMIC_EXCHANGE_POINTER_PTR`?

File: Python/instrumentation.c
Comment (markshannon) [inline Python/instrumentation.c]: Why add the extra argument?
(Same for `initialize_line_tools`)

File: Python/instrumentation.c
Comment (markshannon) [inline Python/instrumentation.c]: Why the atomic operation here?
Won't this leak memory in case of a race?
I think it would be better to lock the code object when doing this initialization.

File: Python/instrumentation.c
Comment (markshannon) [inline Python/instrumentation.c]: Use `ASSERT_WORLD_STOPPED()`, much like we use `ASSERT_WORLD_STOPPED_OR_LOCKED()`  instead of the `#ifdef ...`?

File: Python/instrumentation.c
Comment (markshannon) [inline Python/instrumentation.c]: Rather than three `_PyEval_StartTheWorld(interp);` calls, could you use our standard cleanup idiom of jumping to a `done:` label, and calling `_PyEval_StartTheWorld(interp)` there?

File: Python/instrumentation.c
Comment (markshannon) [reply Python/instrumentation.c:2267]: This is going to be a pain to reason about outside of stop the world. Instrumenting is not thread safe, and invalidates tier 2 traces, which is also not thread safe.

I think we need to make instrumenting code objects safe in a stop the world context. 
There are decrefs when clearing tier 2 executors, but those only reference each other and are safe to deallocate in a stop the world context.

**Review by markshannon — COMMENTED** [review]
I think fine grained locking is the wrong approach here.
PEP 667 specifically states that setting events is expensive.

Stopping the world for setting events and locking code objects to instrument seems like the right approach to me.

**Comment by markshannon** [comment]
Note that it is not possible for a code object that needs to be instrumented to execute without hitting a `RESUME` (or specialized variant) instruction, which will perform the instrumentation.
Provided that `_Py_Instrument` locks the code object, then executing code objects can safely access the instrumentation data structures without locks or atomics.

File: Python/instrumentation.c
Comment (DinoV) [reply Python/instrumentation.c]: Do we really want `_Py_Instrument` to stop the world?  The reason why I didn't want to have that stop the world is because it can happen outside of calls to enable tracing, e.g. in `RESUME` which seems like it would be a massive slow down.

File: Python/instrumentation.c
Comment (DinoV) [reply Python/instrumentation.c]: This was so that we could initialize the data and then publish it so any readers would either see no line data or would see the initialized line data.

File: Python/instrumentation.c
Comment (DinoV) [reply Python/instrumentation.c]: We actually do lock the code object on initialization, but the current implementation was trying to avoid locking the code object on reads.  The readers of `_co_monitoring` are all using acquire semantics so if there is a value there then they'll also pick all of the writes to the data which it is pointing to as well.  The alternative here is to either lock the code object on reads as well or to stop the world on `_Py_Instrument` calls which seems to have disproportionately compared to calls to actually enable monitoring.

**Comment by DinoV** [comment]
> Note that it is not possible for a code object that needs to be instrumented to execute without hitting a `RESUME` (or specialized variant) instruction, which will perform the instrumentation. Provided that `_Py_Instrument` locks the code object, then executing code objects can safely read the instrumentation data structures without locks or atomics.

I don't think locking alone is sufficient, unless you're referring to stopping the world to lock.  If a code object starts running on one thread, hits resume, and another thread enables tracing then calls that function it's going to start mutating the code object.  I'm just concerned that stopping the world to instrument every new code object is going to be extremely expensive, but if that seems like an okay performance cost to you then I'm happy to do that, it is much simpler :)

**Comment by gvanrossum** [comment]
@DinoV Did you actually request a review from me or did I just get that message because I'm a code owner?

File: Lib/test/test_free_threading/test_monitoring.py
Comment (colesbury) [inline Lib/test/test_free_threading/test_monitoring.py]:
```suggestion
    def test_instrumentation(self):
```

**Comment by DinoV** [comment]
> @DinoV Did you actually request a review from me or did I just get that message because I'm a code owner?

I think you just got it because you're a code owner.

**Comment by DinoV** [comment]
@markshannon @brandtbucher This adds some new parallel monitoring tests which @colesbury figured out crash even w/o any changes to the monitoring infrastructure.

**Comment by DinoV** [comment]
> @markshannon @brandtbucher This adds some new parallel monitoring tests which @colesbury figured out crash even w/o any changes to the monitoring infrastructure.

So the issue here is that the `CHECK_EVAL_BREAKER();` in `ENTER_EXECUTOR` (https://github.com/python/cpython/blob/main/Python/bytecodes.c#L2396) can release the GIL, allow another thread to run, and that thread can clear the `co_executors` on the code object by enabling instrumentation.

**Comment by brandtbucher** [comment]
Ah, interesting. Thanks for digging that up.

Easy solution would be to move the `CHECK_EVAL_BREAKER();` after the `Py_INCREF(executor);` at the end with a comment. Arguably more "correct" solution would be to check the opcode after the `CHECK_EVAL_BREAKER();` and re-run the instruction if it's anything other than `ENTER_EXECUTOR`.

I'm slightly leaning towards the former, since the tier two code should bail immediately anyways upon seeing it's been invalidated, and it keeps the redundant check for the opcode out of the happy path.

**Comment by brandtbucher** [comment]
Hm, actually, I think that could leak the incref'ed executor if the eval breaker check raises (I forgot there's a `goto error` hidden in that macro).

File: Python/bytecodes.c
Comment (gvanrossum) [inline Python/bytecodes.c]: This concerns me a bit. `CHECK_EVAL_BREAKER()` contains a `goto error` and I'm worried that that would leak the executor.

File: Include/cpython/pyatomic_msc.h
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```suggestion
#  error "no implementation of _Py_atomic_load_uintptr_acquire"
```

File: Include/cpython/pyatomic_msc.h
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```suggestion
#  error "no implementation of _Py_atomic_store_uintptr_release"
```

File: Python/instrumentation.c
Comment (colesbury) [inline Python/instrumentation.c]: Should be `#define UNLOCK_CODE()` without an argument

File: Python/legacy_tracing.c
Comment (colesbury) [inline Python/legacy_tracing.c]: GitHub warnings about `Py_ssize_t` vs. `int`.

**Review by colesbury — COMMENTED** [review]
A few random things I noticed while reading the PR

File: Python/bytecodes.c
Comment (gvanrossum) [reply Python/bytecodes.c]: Thanks for fixing this!

**Comment by DinoV** [comment]
> Hm, actually, I think that could leak the incref'ed executor if the eval breaker check raises (I forgot there's a `goto error` hidden in that macro).

Okay, i've gone with re-checking the opcode approach.  It looks like in addition to checking the opcode we've also got to check the oparg as we can seemingly end up just updating the oparg.

**Review by gvanrossum — COMMENTED** [review]
(You're not waiting for me are you?)

**Comment by DinoV** [comment]
> (You're not waiting for me are you?)

Nope, I'm more hoping to get a review from @markshannon and/or @colesbury unless you feel particularly familiar with this area :)

File: Python/instrumentation.c
Comment (colesbury) [inline Python/instrumentation.c]: I feel like we are taking the `FT_` macros too far. I think just using a `_Py_atomic_exchange_ptr` is going to be easier to understand and maintain.

There are some contexts where we don't want to pay for the cost of an atomic op in the default build, but I don't think it matters here.

File: Python/instrumentation.c
Comment (colesbury) [reply Python/instrumentation.c:2267]: I agree that we want to do the instrumentation in a stop the world context. The only thing I'd consider is refactoring so that the `Py_DECREF`s happen later after the world is started again. But if the executors are guaranteed to have simple destructors that won't call back into Python than it's certainly a lot easier to keep things structured the way they are now.

**Review by colesbury — APPROVED** [review]
LGTM!

**Comment by DinoV** [comment]
@markshannon Do you want to take a look at this or are you happy with where it's at?

**Pull request merged**
